python/lsst/daf/butler/registry/_registry_factory.py

```diff
@@ -0,0 +1,185 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("RegistryFactory",)
+
+from typing import TYPE_CHECKING
+
+from lsst.resources import ResourcePathExpression
+from lsst.utils import doImportType
+
+from ..core import Config, DimensionConfig
+from ._butler_registry import ButlerRegistry
+from ._config import RegistryConfig
+from ._defaults import RegistryDefaults
+
+if TYPE_CHECKING:
+    from .._butlerConfig import ButlerConfig
+
+
+class RegistryFactory:
+    """Interface for creating and initializing Registry instances.
+
+    Each registry implementation can have its own constructor parameters.
+    The assumption is that an instance of a specific subclass will be
+    constructed from configuration using ``RegistryClass.fromConfig()`` or
+    ``RegistryClass.createFromConfig()``.
+
+    This class will look for a ``cls`` entry in registry configuration object
+    (defaulting to ``SqlRegistry``), import that class, and call one of the
+    above methods on the imported class.
+    """
+
```

Comment:
Would it make sense to give this class a constructor that does the work of `force_registry_config` and holds the config as an attribute, and then make the other methods regular instance methods (probably abstract ones)?

Reply:
Adding a constructor looks reasonable. If factory methods are abstract we'd need them implemented somewhere? Do you think that this should be another base class for ButlerRegistry subclasses?

Reply:
Ah, I think I got confused about the current state of things while reviewing commit-by-commit - I thought this still had some methods with docs that said they needed to be reimplemented, and that made me think there were _already_ subclass implementations and I didn't step back and notice that there weren't any.

So just ignore the bit about abstract methods; this is just a regular concrete (probably final) class, and that's fine.